lib/srv/app/server.go

```diff
@@ -0,0 +1,501 @@
+		return trace.BadParameter("heartbeat missing")
+	}
+
+	return nil
+}
+
+// Server is an application server. It authenticates requests from the web
+// proxy and forwards them to internal applications.
+type Server struct {
+	c   *Config
+	log *logrus.Entry
+
+	closeContext context.Context
+	closeFunc    context.CancelFunc
+
+	httpServer *http.Server
+	tlsConfig  *tls.Config
+
+	heartbeat     *srv.Heartbeat
+	dynamicLabels map[string]*labels.Dynamic
+
+	keepAlive time.Duration
+	proxyPort string
+
+	cache *sessionCache
+}
+
+// New returns a new application server.
+func New(ctx context.Context, c *Config) (*Server, error) {
+	err := c.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Server{
+		c: c,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: teleport.ComponentApp,
+		}),
+	}
+
+	s.closeContext, s.closeFunc = context.WithCancel(ctx)
+
+	// Make copy of servers TLS configuration and update it with the specific
+	// functionality this server needs, like requiring client certificates.
+	s.tlsConfig = copyAndConfigureTLS(s.c.TLSConfig, s.getConfigForClient)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create and configure HTTP server with authorizing middleware.
+	s.httpServer = s.newHTTPServer()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a new session cache, this holds sessions that can be used to
+	// forward requests.
+	s.cache, err = newSessionCache(s.closeContext, s.log)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create dynamic labels for all applications that are being proxied and
+	// sync them right away so the first heartbeat has correct dynamic labels.
+	s.dynamicLabels = make(map[string]*labels.Dynamic)
+	for _, a := range c.Server.GetApps() {
+		if len(a.DynamicLabels) == 0 {
+			continue
+		}
+		dl, err := labels.NewDynamic(s.closeContext, &labels.DynamicConfig{
+			Labels: services.V2ToLabels(a.DynamicLabels),
+			Log:    s.log,
+		})
+		if err != nil {
+
+			return nil, trace.Wrap(err)
+		}
+		dl.Sync()
+		s.dynamicLabels[a.Name] = dl
+	}
+
+	// Create heartbeat loop so applications keep sending presence to backend.
+	s.heartbeat, err = srv.NewHeartbeat(srv.HeartbeatConfig{
+		Mode:            srv.HeartbeatModeApp,
+		Context:         s.closeContext,
+		Component:       teleport.ComponentApp,
+		Announcer:       c.AccessPoint,
+		GetServerInfo:   s.GetServerInfo,
+		KeepAlivePeriod: defaults.ServerKeepAliveTTL,
+		AnnouncePeriod:  defaults.ServerAnnounceTTL/2 + utils.RandomDuration(defaults.ServerAnnounceTTL/2),
+		CheckPeriod:     defaults.HeartbeatCheckPeriod,
+		ServerTTL:       defaults.ServerAnnounceTTL,
+		OnHeartbeat:     c.OnHeartbeat,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Pick up TCP keep-alive settings from the cluster level.
+	clusterConfig, err := s.c.AccessPoint.GetClusterConfig()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	s.keepAlive = clusterConfig.GetKeepAliveInterval()
```

Comment:
also set the `KeepAlivePeriod` in `Heartbeat` from this

Reply:
That period is used to also control how much bandwidth the cluster uses when heartbeating presence information. Not sure if this is the place to allow control of it.